platform/ios/Integration Tests/Snapshotter Tests/MGLMapSnapshotterTest.m

```diff
@@ -121,41 +234,25 @@ - (void)testSnapshotterFromBackgroundQueue {
     dispatch_queue_attr_t attr = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_USER_INITIATED, QOS_MIN_RELATIVE_PRIORITY); // also for concurrent
     dispatch_queue_t backgroundQueue = dispatch_queue_create(__PRETTY_FUNCTION__, attr);
 
-
     // Use dispatch_group to keep the backgroundQueue block around (and
     // so also the MGLMapSnapshotter
     dispatch_group_t group = dispatch_group_create();
     dispatch_group_enter(group);
 
-
     dispatch_async(backgroundQueue, ^{
 
         MGLMapSnapshotter *snapshotter = snapshotterWithCoordinates(coord, size);
         XCTAssertNotNil(snapshotter);
 
         MGLMapSnapshotCompletionHandler completion = ^(MGLMapSnapshot * _Nullable snapshot, NSError * _Nullable error) {
-
-            // This should be the main queue
-            __typeof__(self) strongself = weakself;
-
-            MGLTestAssertNotNil(strongself, strongself);
-
-            MGLTestAssertNotNil(strongself, snapshot);
-            MGLTestAssertNotNil(strongself, snapshot.image);
-            MGLTestAssertNil(strongself, error, @"Snapshot should not error with: %@", error);
-
-            // Change this to XCTAttachmentLifetimeKeepAlways to be able to look at the snapshots after running
-            XCTAttachment *attachment = [XCTAttachment attachmentWithImage:snapshot.image];
-            attachment.lifetime = XCTAttachmentLifetimeDeleteOnSuccess;
-            [strongself addAttachment:attachment];
-
+            // The completion block should not be called
+            MGLTestFail(weakself);
             dispatch_group_leave(group);
         };
 
-        // untested
         @try {
             [snapshotter startWithCompletionHandler:completion];
-            MGLTestFail(weakself);
+            MGLTestFail(weakself, @"startWithCompletionHandler: should raise an exception");
```

Comment:
Specifically, the requirement here is that you can't start the snapshotter from a thread that doesn't have its own RunLoop, right?  Thinking back to https://github.com/mapbox/mapbox-gl-native/issues/12336#issuecomment-403217950, is there a way we can support starting from an arbitrary dispatch queue by allowing the user to provide an existing `RunLoop`. What I'm thinking of is how some of the complexity for the OP was due to calling the code from an async dispatch block on the main queue -- when really the only reason it had to run on the main queue was to get `mbgl::Scheduler::GetCurrent()`, right?

Reply:
When I looked back at this code it was definitely confusing as to what was going on, which is why I updated the "assert" comment above. 

> is there a way we can support starting from an arbitrary dispatch queue by allowing the user to provide an existing RunLoop

This is getting out of my knowledge area 😄 but I agree that providing such a mechanism would be handy, as currently we can only ever call `start..` from the main queue. What's the process of creating a `RunLoop` for a dispatch queue?

Reply:
🤔 OK I recognize this is a lot of scope expansion, but I think it might be worth taking this on while we're motivated by the original bug report and we've got our heads wrapped around what we expect from these tests.

Thinking about it a little more, I don't think we really want to create a `RunLoop` on the dispatch queue (although you could). The current `RunLoop` is just what we use to get a `Scheduler` to pass to the `MapSnapshotter::Callback`. It needs someone to process the callback completion message -- but I don't think there's actually a requirement that the callback be processed on the main/UI thread. Instead of passing `*mbgl::Scheduler::GetCurrent()`, we could pass `mbgl::sharedThreadPool()`, so the completion would be handled on one the worker threads we already have set up. This would also remove the necessity of the extra step where we do `drawAttributedSnapshotWorker` on a separate worker queue, since we'd already be on a worker (albeit the mbgl shared worker pool instead of a GCD worker queue).

The complication here is that we need to make sure we don't modify the `MGLMapSnapshotter` object (aka `strongSelf`) from a thread that's different from the one that owns it. That was effectively accomplished before by forcing the callback to run on the main thread. Most of the isolation we need is already accomplished by `drawAttributedWorkerSnapshot`, we just have to move the capture of things like `strongSelf.options.scale` up another level.

The tricky part is handling the logic for the destruction of the `_snapshotCallback` (and allowing the calling thread to create a new one). I think you can do the destruction of `_snapshotCallback` from the background, but do it _before_ you call the completion handler, and _before_ you set `strongSelf.loading = false`. You could set `loading` to be atomic, too. That way, you could be guaranteed that whoever the calling thread was, it wouldn't try to create a new callback until the old one was torn down. Also we want to make sure we're done with the tear down _before_ we queue the completion callback, because the completion callback is likely to run on the thread that owns the `MGLMapSnapshotter`, and whoever receives the completion is likely to assume they can immediately queue another request.